### ISSUE TITLE
Something leaves classes.jar files open until MSBuild exits or GC runs finalizers

### ISSUE THREAD

**Issue by KalleOlaviNiemitalo** [xamarin/xamarin-android#5424]
### Steps to Reproduce

A hand-crafted MSBuild project builds several variants of a Xamarin Android app. For each variant, it edits source files, deletes the `obj` and `bin` directories with the RemoveDir task, builds the app with the MSBuild task (`Targets="SignAndroidPackage"`), and copies the resulting APK to another directory.

See <https://github.com/xamarin/xamarin-android/issues/5424#issuecomment-748673702> for a demo.

### Expected Behavior

All builds succeed.

### Actual Behavior

The RemoveDir task fails because it cannot delete `obj\Release\90\lp\46\jl\bin\classes.jar` and `obj\Release\90\lp\52\jl\bin\classes.jar`. This error does not happen with the variant that is built first. Process Monitor shows that MSBuild.exe closes those files after the error.

### Version Information

The problem has been happening with Visual Studio Build Tools 2017 v15.9.30 on Windows Server 2012 R2. That does not have an IDE, so I cannot check the Xamarin version from there. However, the corresponding version of Visual Studio Professional 2017 lists:

Microsoft Visual Studio Professional 2017 
Version 15.9.30
VisualStudio.15.Release/15.9.30+28307.1321
Microsoft .NET Framework
Version 4.8.04084

Installed Version: Professional

Microsoft Visual Studio Tools for Applications 2017   00370-20007-48825-AA894
Microsoft Visual Studio Tools for Applications 2017

Visual C++ 2017   00370-20007-48825-AA894
Microsoft Visual C++ 2017

.NET Portability Analyzer   1.1.10808.0
Evaluates portability of assemblies across .NET platforms.

Application Insights Tools for Visual Studio Package   8.14.20131.1
Application Insights Tools for Visual Studio

ASP.NET and Web Tools 2017   15.9.04012.0
ASP.NET and Web Tools 2017

ASP.NET Core Razor Language Services   15.8.31590
Provides languages services for ASP.NET Core Razor.

ASP.NET Web Frameworks and Tools 2017   5.2.61435.0
For additional information, visit https://www.asp.net/

Azure App Service Tools v3.0.0   15.9.03024.0
Azure App Service Tools v3.0.0

C# Tools   2.10.0-beta2-63501-03+b9fb1610c87cccc8ceb74a770dba261a58e39c4a
C# components used in the IDE. Depending on your project type and settings, a different version of the compiler may be used.

Common Azure Tools   1.10
Provides common services for use by Azure Mobile Services and Microsoft Azure Tools.

Extensibility Message Bus   1.1.49 (remotes/origin/d15-8@ee674f3)
Provides common messaging-based MEF services for loosely coupled Visual Studio extension components communication and integration.

JavaScript Language Service   2.0
JavaScript Language Service

JavaScript Project System   2.0
JavaScript Project System

JavaScript UWP Project System   2.0
JavaScript UWP Project System

Microsoft Continuous Delivery Tools for Visual Studio   0.4
Simplifying the configuration of Azure DevOps pipelines from within the Visual Studio IDE.

Microsoft JVM Debugger   1.0
Provides support for connecting the Visual Studio debugger to JDWP compatible Java Virtual Machines

Microsoft Library Manager   1.0
Install client-side libraries easily to any web project

Microsoft MI-Based Debugger   1.0
Provides support for connecting Visual Studio to MI compatible debuggers

Microsoft Visual C++ Wizards   1.0
Microsoft Visual C++ Wizards

Microsoft Visual Studio Tools for Containers   1.1
Develop, run, validate your ASP.NET Core applications in the target environment. F5 your application directly into a container with debugging, or CTRL + F5 to edit & refresh your app without having to rebuild the container.

Microsoft Visual Studio VC Package   1.0
Microsoft Visual Studio VC Package

MLGen Package Extension   1.0
MLGen Package Visual Studio Extension Detailed Info

Mono Debugging for Visual Studio   4.13.12-pre (9bc9548)
Support for debugging Mono processes with Visual Studio.

NuGet Package Manager   4.6.0
NuGet Package Manager in Visual Studio. For more information about NuGet, visit http://docs.nuget.org/.

ProjectServicesPackage Extension   1.0
ProjectServicesPackage Visual Studio Extension Detailed Info

ResourcePackage Extension   1.0
ResourcePackage Visual Studio Extension Detailed Info

ResourcePackage Extension   1.0
ResourcePackage Visual Studio Extension Detailed Info

SQL Server Data Tools   15.1.62002.01090
Microsoft SQL Server Data Tools

Syntax Visualizer   1.0
An extension for visualizing Roslyn SyntaxTrees.

TypeScript Tools   15.9.30718.2001
TypeScript Tools for Microsoft Visual Studio

Visual Basic Tools   2.10.0-beta2-63501-03+b9fb1610c87cccc8ceb74a770dba261a58e39c4a
Visual Basic components used in the IDE. Depending on your project type and settings, a different version of the compiler may be used.

Visual F# Tools 10.2 for F# 4.5   15.8.0.0.  Commit Hash: 6e26c5bacc8c4201e962f5bdde0a177f82f88691.
Microsoft Visual F# Tools 10.2 for F# 4.5

Visual Studio Code Debug Adapter Host Package   1.0
Interop layer for hosting Visual Studio Code debug adapters in Visual Studio

Visual Studio Tools for Containers   1.0
Visual Studio Tools for Containers

Visual Studio Tools for Universal Windows Apps   15.0.28307.1300
The Visual Studio Tools for Universal Windows apps allow you to build a single universal app experience that can reach every device running Windows 10: phone, tablet, PC, and more. It includes the Microsoft Windows 10 Software Development Kit.

VisualStudio.Mac   1.0
Mac Extension for Visual Studio

Xamarin   4.12.3.83 (d15-9@23fa80172)
Visual Studio extension to enable development for Xamarin.iOS and Xamarin.Android.

Xamarin Designer   4.16.30 (e0af99a34)
Visual Studio extension to enable Xamarin Designer tools in Visual Studio.

Xamarin Templates   1.1.128 (6f5ebb2)
Templates for building iOS, Android, and Windows apps with Xamarin and Xamarin.Forms.

Xamarin.Android SDK   9.1.7.0 (HEAD/ba9da7a76)
Xamarin.Android Reference Assemblies and MSBuild support.

Xamarin.iOS and Xamarin.Mac SDK   12.4.0.64 (9c8d8e0)
Xamarin.iOS and Xamarin.Mac Reference Assemblies and MSBuild support.

### Workarounds

I have been able to work around the problem in two ways:

1. By running MSBuild.exe with the Exec task, instead of using the MSBuild task. However, this makes MSBuild logging more difficult to control.
2. By making a custom task call GC.Collect() and GC.WaitForPendingFinalizers() after the MSBuild task finishes. This seems the better workaround.

### Suspected cause

I have not been able to verify which MSBuild task leaves the classes.jar files open.

I suspect the `ZipArchive.Open (File.OpenRead (jarFile))` call in the BuildApk task: <https://github.com/xamarin/xamarin-android/blob/b0283cf5523c08139fc77c5ab59b5444b51a0e5f/src/Xamarin.Android.Build.Tasks/Tasks/BuildApk.cs#L173>

As far as I can tell, [ZipArchive.Open](https://github.com/xamarin/LibZipSharp/blob/8f2053c7d6d7108ed49ca88aab9ef5134929eddc/ZipArchive.cs#L170) returns a ZipArchive whose Dispose method never closes the Stream. ~~However, the [ZipArchive constructor](https://github.com/xamarin/LibZipSharp/blob/8f2053c7d6d7108ed49ca88aab9ef5134929eddc/ZipArchive.cs#L83) also seems to leak a GCHandle to the Stream ([LibZipSharp never calls GCHandle.Free](https://github.com/xamarin/LibZipSharp/search?q=%22GCHandle.Free%22)), so if this File.OpenRead were the problem, then GC.Collect() and GC.WaitForPendingFinalizers() would not have helped because the finalizer would not have run. Unless perhaps I am looking at the wrong version of LibZipSharp.~~

**Comment by jonathanpeppers** [comment]
@dellis1972 / @grendello you might be interested in the `GCHandle.Free()` comment ^^

@KalleOlaviNiemitalo can you share a [diagnostic MSBuild log](https://docs.microsoft.com/xamarin/android/troubleshooting/troubleshooting#diagnostic-msbuild-output) in case it's something else?

**Comment by KalleOlaviNiemitalo** [comment]
I was mistaken about GCHandle. There is no GCHandle leak in LibZipSharp; the GCHandle.Free call is at <https://github.com/xamarin/LibZipSharp/blob/8f2053c7d6d7108ed49ca88aab9ef5134929eddc/ZipArchive.cs#L821-L824>.

Because of the GCHandle misunderstanding, I had thought that the fault could not be in the `ZipArchive.Open (File.OpenRead (jarFile))` call in the BuildApk task. I now believe that is the cause after all, based on poking with Process Monitor and SOS in WinDbg.

I set up a demo project without proprietary code: [FileHandle5424.zip](https://github.com/xamarin/xamarin-android/files/5721530/FileHandle5424.zip). This includes an MSBuild binary log, built like so:

1. Start Command Prompt
2. Delete some environment variables that I did not want to disclose
3. Run `"C:\Program Files (x86)\Microsoft Visual Studio\2017\Professional\Common7\Tools\VsDevCmd.bat" -arch=amd64`
4. Run `nuget restore FileHandle5424.sln` with NuGet Version: 5.4.0.6315
5. Run `"C:\Program Files (x86)\Microsoft Visual Studio\2017\Professional\MSBuild\15.0\Bin\amd64\MSBuild.exe" /binaryLogger:MSBuild.binlog;ProjectImports=None Demo.msbuildproj`

The issue occurs more reliably with a 64-bit MSBuild.exe process. A 32-bit process collects garbage too often and closes the files at that time.

It seems this issue mainly applies to projects using `packages.config`. If the project were using `PackageReference` instead, then deleting the `obj` directory after building the first variant would make the second variant fail to build in any case.
